blockchain/utxocache.go

```diff
@@ -0,0 +1,790 @@
+	totalEntrySize uint64
+
+	// The following fields track the total number of cache hits and misses and
+	// are used to measure the overall cache hit ratio.
+	hits   uint64
+	misses uint64
+}
+
+// UtxoCacheConfig is a descriptor which specifies the utxo cache instance
+// configuration.
+type UtxoCacheConfig struct {
+	// DB defines the database which houses the utxo set.
+	//
+	// This field is required.
+	DB database.DB
+
+	// MaxSize defines the maximum allowed size of the utxo cache, in bytes.
+	//
+	// This field is required.
+	MaxSize uint64
+}
+
+// NewUtxoCache returns a UtxoCache instance using the provided configuration
+// details.
+func NewUtxoCache(config *UtxoCacheConfig) *UtxoCache {
+	// Approximate the maximum number of entries allowed in the cache in order
+	// to set the initial capacity of the entries map.
+	avgEntrySize := outpointSize + pointerSize + baseEntrySize + p2pkhScriptLen
+	maxEntries := math.Ceil(float64(config.MaxSize) / float64(avgEntrySize))
+
+	return &UtxoCache{
+		db:            config.DB,
+		maxSize:       config.MaxSize,
+		entries:       make(map[wire.OutPoint]*UtxoEntry, uint64(maxEntries)),
+		lastFlushTime: time.Now(),
+	}
+}
+
+// totalSize returns the total size of the cache on a 64-bit platform, in bytes.
+// Note that this only takes the entries map into account, which represents the
+// vast majoirty of the memory that the cache uses, and does not include the
+// memory usage of other fields in the utxo cache struct.
+//
+// This function MUST be called with the cache lock held.
+func (c *UtxoCache) totalSize() uint64 {
+	numEntries := uint64(len(c.entries))
+	return outpointSize*numEntries + pointerSize*numEntries + c.totalEntrySize
+}
+
+// hitRatio returns the percentage of cache lookups that resulted in a cache
+// hit.
+//
+// This function MUST be called with the cache lock held.
+func (c *UtxoCache) hitRatio() float64 {
+	totalLookups := c.hits + c.misses
+	if totalLookups == 0 {
+		return 100
+	}
+
+	return float64(c.hits) / float64(totalLookups) * 100
+}
+
+// addEntry adds the specified output to the cache.  The entry being added MUST
+// NOT be mutated by the caller after being passed to this function.
+//
+// Note that this function does not check if the entry is unspendable and
+// therefore the caller should ensure that the entry is spendable before adding
+// it to the cache.
+//
+// This function MUST be called with the cache lock held.
+func (c *UtxoCache) addEntry(outpoint wire.OutPoint, entry *UtxoEntry) error {
+	// Attempt to get an existing entry from the cache.
+	cachedEntry := c.entries[outpoint]
+
+	// If an existing entry does not exist, the added entry should be marked as
+	// fresh.
+	if cachedEntry == nil {
+		entry.state |= utxoStateFresh
+	}
+
+	// Entries being added to the cache should always be marked as modified to
+	// ensure that they are flushed to the database.
+	entry.state |= utxoStateModified
```

Comment:
> Perhaps there is something later that makes this always the case, but I don't see really why this should always be true. You could load up an entry that is only used for the purposes of providing data to the build the transaction index, for example, in which case the utxo would never be modified.

In this case, `addEntry` would not called, so we wouldn't unnecessarily write entries back to the database.

> I don't think it would really hurt anything to be paranoid

This is mainly why I had set it in case things changed in the future.  As I iterated the solution, though, the only place that `addEntry` is actually called is when [committing the view to the cache here](https://github.com/decred/dcrd/blob/80a14d1c800e420b0d07bd3fbc8674fd7a1e7287/blockchain/utxocache.go#L385-L387), so I agree that this can safely be removed and just set when marking the entry as fresh.

Reply:
Updated this to only mark as modified when marking the entry as fresh.